---
title: "Clustering samples"
author: "Bruno de Medeiros"
date: "7/16/2018"
output: 
  html_notebook:
    df_print: paged
---

We used PCANGSD to obtain variance-covariance matrices for our samples. Now we will use these to get their eigenvalues and eigenvectors, and use k-means to cluster. This is similar to what R package adegenet does (we will steal some commands from find.clusters), but by using ANGSD we are able to include loci sequenced at low coverage and imputing for missign data is not simply the average allele frequencies, but it is informed by the other alleles as well.

# Load packages

Let's start by reading packages:
```{r warning=FALSE, message=FALSE}
library(raster)
library(rnaturalearth)
library(sf)
library(adegenet)
library(tidyverse)
library(ggthemes)
library(RStoolbox)
library(scales)
library(xtable)
library(ggrepel)
rm(list=ls())
```

# Load sample ids

PCANGSD does not list sample names in the variance/covariance matrix, so we will have to read these from the list of bam files passed to ANGSD first.

```{r}
spp = list.dirs(file.path('..','ANGSD_weevils/PCA'),recursive = F,full.names = F)

names(spp) = str_c(spp,'_')

insect_specimens = spp %>%
  sapply(function(x){
    list.files(file.path('..','ANGSD_weevils/PCA',x), 
               pattern = '^bam.filelist',
               recursive = T,
               full.names = T) %>%
      file() %>%
      readLines() %>%
      basename() %>%
      str_remove_all('\\..+$')
  }) %>%
  unlist

sample_names = tibble(sample = insect_specimens, 
                      taxon = names(insect_specimens) %>% str_remove('_[0-9]+$'))
sample_names

#we will use this later
names(spp) = spp

```

# Compute eigenvalues and eigenvectors

Let's now load variance/covariance matrices
```{r}
covs = spp %>%
  plyr::llply(function(x){
    list.files(file.path('..','ANGSD_weevils/PCA',x), 
               pattern = '.+pca\\.cov',
               recursive = T,
               full.names = T) %>%
      file %>%
      readr::read_lines() %>%
      plyr::laply(str_split,'[:space:]',simplify = T) %>%
      as.numeric %>%
      matrix(.,nrow=sqrt(length(.)))
  })

str(covs)
```

For each matrix, let's compute the eigenvalues and eigenvectors.


```{r}
eig = plyr::llply(covs, eigen, symmetric = T)
str(eig)
```


# K-means clustering

We will now use a modified version of adegenet's function `find.clusters` to find the optimal number of clusters for each species. Let's define the function first:

```{r}
find.clusters.eigen = function (x, n.pca = NULL, n.clust = NULL, method = c("kmeans", 
    "ward"), stat = c("BIC", "AIC", "WSS"), choose.n.clust = TRUE, 
    criterion = c("diffNgroup", "min", "goesup", "smoothNgoesup", 
        "goodfit"), max.n.clust = round(nrow(x)/10), n.iter = 1e+05, 
    n.start = 10, center = TRUE, scale = TRUE, pca.select = c("nbEig", 
        "percVar"), perc.pca = NULL, ..., dudi = NULL) 
{
    .compute.wss = adegenet:::.compute.wss
    stat <- match.arg(stat)
    pca.select <- match.arg(pca.select)
    criterion <- match.arg(criterion)
    min.n.clust <- 2
    max.n.clust <- max(max.n.clust, 2)
    method <- match.arg(method)
    
    

    N <- nrow(x$vectors)
    # REDUCEDIM <- is.null(dudi)
    # if (REDUCEDIM) {
    #     maxRank <- min(dim(x))
    #     pcaX <- dudi.pca(x, center = center, scale = scale, scannf = FALSE, 
    #         nf = maxRank)
    # }
    # else {
    #     pcaX <- dudi
    # }
    cumVar <- 100 * cumsum(x$values)/sum(x$values)
    #eigenvectors are normalized, adegenet expects them multiplied by sqrt(eigenvalue)
    x$vectors = apply(x$vectors,1,function(y) y * sqrt(x$values)) %>% t 
    # if (!REDUCEDIM) {
    #     myCol <- rep(c("black", "lightgrey"), c(ncol(pcaX$li), 
    #         length(pcaX$eig)))
    # }
    # else {
        myCol <- "black"
    # }
    if (is.null(n.pca) & pca.select == "nbEig") {
        plot(cumVar, xlab = "Number of retained PCs", ylab = "Cumulative variance (%)", 
            main = "Variance explained by PCA", col = myCol)
        cat("Choose the number PCs to retain (>= 1): ")
        n.pca <- as.integer(readLines(con = getOption("adegenet.testcon"), 
            n = 1))
    }
    if (is.null(perc.pca) & pca.select == "percVar") {
        plot(cumVar, xlab = "Number of retained PCs", ylab = "Cumulative variance (%)", 
            main = "Variance explained by PCA", col = myCol)
        cat("Choose the percentage of variance to retain (0-100): ")
        nperc.pca <- as.numeric(readLines(con = getOption("adegenet.testcon"), 
            n = 1))
    }
    if (!is.null(perc.pca)) {
        n.pca <- min(which(cumVar >= perc.pca))
        if (perc.pca > 99.999) 
            n.pca <- length(pcaX$eig)
        if (n.pca < 1) 
            n.pca <- 1
    }
    X.rank <- length(x$values)
    n.pca <- min(X.rank, n.pca)
    if (n.pca >= N) 
        warning("number of retained PCs of PCA is greater than N")
    XU <- x$vectors[, 1:n.pca, drop = FALSE]
    if (is.null(n.clust)) {
        nbClust <- min.n.clust:max.n.clust
        WSS <- numeric(0)
        for (i in 1:length(nbClust)) {
            if (method == "kmeans") {
                temp <- kmeans(XU, centers = nbClust[i], iter.max = n.iter, 
                  nstart = n.start)
            }
            else {
                temp <- list()
                temp$cluster <- cutree(hclust(dist(XU)^2, method = "ward.D2"), 
                  k = nbClust[i])
            }
            WSS[i] <- .compute.wss(XU, temp$cluster)
        }
        if (stat == "AIC") {
            WSS.ori <- sum(apply(XU, 2, function(v) sum((v - 
                mean(v))^2)))
            k <- nbClust
            myStat <- N * log(c(WSS.ori, WSS)/N) + 2 * c(1, nbClust)
            myLab <- "AIC"
            myTitle <- "Value of AIC \nversus number of clusters"
        }
        if (stat == "BIC") {
            WSS.ori <- sum(apply(XU, 2, function(v) sum((v - 
                mean(v))^2)))
            k <- nbClust
            myStat <- N * log(c(WSS.ori, WSS)/N) + log(N) * c(1, 
                nbClust)
            myLab <- "BIC"
            myTitle <- "Value of BIC \nversus number of clusters"
        }
        if (stat == "WSS") {
            WSS.ori <- sum(apply(XU, 2, function(v) sum((v - 
                mean(v))^2)))
            myStat <- c(WSS.ori, WSS)
            myLab <- "Within sum of squares"
            myTitle <- "Value of within SS\nversus number of clusters"
        }
        if (choose.n.clust) {
            plot(c(1, nbClust), myStat, xlab = "Number of clusters", 
                ylab = myLab, main = myTitle, type = "o", col = "blue")
            abline(h = 0, lty = 2, col = "red")
            cat("Choose the number of clusters (>=2: ")
            n.clust <- NA
            while (is.na(n.clust)) {
                n.clust <- max(1, as.integer(readLines(con = getOption("adegenet.testcon"), 
                  n = 1)))
            }
        }
        else {
            if (criterion == "min") {
                n.clust <- which.min(myStat)
            }
            if (criterion == "goesup") {
                n.clust <- min(which(diff(myStat) > 0))
            }
            if (criterion == "goodfit") {
                temp <- min(myStat) + 0.1 * (max(myStat) - min(myStat))
                n.clust <- min(which(myStat < temp)) - 1
            }
            if (criterion == "diffNgroup") {
                temp <- cutree(hclust(dist(diff(myStat)), method = "ward.D"), 
                  k = 2)
                goodgrp <- which.min(tapply(diff(myStat), temp, 
                  mean))
                n.clust <- max(which(temp == goodgrp)) + 1
            }
            if (criterion == "smoothNgoesup") {
                temp <- myStat
                temp[2:(length(myStat) - 1)] <- sapply(1:(length(myStat) - 
                  2), function(i) mean(myStat[c(i, i + 1, i + 
                  2)]))
                n.clust <- min(which(diff(temp) > 0))
            }
        }
    }
    else {
        myStat <- NULL
    }
    if (n.clust > 1) {
        if (method == "kmeans") {
            best <- kmeans(XU, centers = n.clust, iter.max = n.iter, 
                nstart = n.start)
        }
        else {
            best <- list()
            best$cluster <- cutree(hclust(dist(XU)^2, method = "ward.D2"), 
                k = n.clust)
            best$size <- table(best$cluster)
        }
    }
    else {
        best <- list(cluster = factor(rep(1, N)), size = N)
    }
    if (!is.null(myStat)) {
        names(myStat) <- paste("K", c(1, nbClust), sep = "=")
    }
    res <- list(Kstat = myStat, stat = myStat[n.clust], grp = factor(best$cluster), 
        size = best$size)
    return(res)
}
```



Now let's loop through species and use find.clusters to identify the optimal number of clusters.
We used the following code to choose the numbers interactively, preserving using the PCA axes corresponding to 80% of the genetic variance. The code is commented out because it is run interactively only once


```{r}
clusters = spp %>%
  plyr::llply(function(x){
    find.clusters.eigen(eig[[x]],
                  pca.select = 'percVar',
                  perc.pca=90,
                  max.n.clust = min(10, length(eig[[x]]$values)%/%2))
  },
  .parallel = FALSE)

save(clusters,file = 'temporary_data/clusters_DAPC.RData')
```

Let's load the object above if this is a re-run of this notebook. It turns out that adegenet does not include sample names if there is only one cluster, so we have to standardize that before proceeding

```{r}
load('temporary_data/clusters_DAPC.RData')
clusters = spp %>%
  plyr::llply(function(i_sp){
    result = clusters[[i_sp]]
    if(is.null(names(result$grp))){
      names(result$grp) = sample_names %>% filter(taxon == i_sp) %>% pull(sample)
    }
    return(result)
  })
str(clusters,max.level = 1)
```

Now that we have clusters, let's visualize them in a PCA plot. For that, let's first create a data.frame including information about species, group assignment and PCA axes 1-2
```{r warning=FALSE, message=FALSE}

PCA_df = spp %>%
  plyr::ldply(function(i_sp){
    PCA_axes = eig[[i_sp]]$vectors[,1:2] %>%
      as.data.frame %>%
      rename(PC1 = 1, PC2 = 2) %>%
      mutate(PC1 = PC1 * sqrt(eig[[i_sp]]$values[1]),
             PC2 = PC2 * sqrt(eig[[i_sp]]$values[2]),
             cluster = clusters[[i_sp]]$grp) %>%
      bind_cols(filter(sample_names, taxon == i_sp)) %>%
      dplyr::select(-taxon) #will be added automatically
  },
  .parallel = FALSE) %>%
  rename(taxon = .id)

PCA_df
```

Before plotting, let's create a vector to produce nice species names in the output:
```{r}
names_translation = c('Anchylorhynchus_1' = "italic(Anc.~trapezicollis)~OTU~1",
                     'Anchylorhynchus_2' = "italic(Anc.~trapezicollis)~OTU~2",
                     'Anchylorhynchus_3' = "italic(Anc.~trapezicollis)~OTU~3",
                     'Andranthobius_1' = "italic(And.~bondari)~OTU~1",
                     'Andranthobius_2' = "italic(And.~bondari)~OTU~2",
                     'C_decolor_1' = "italic(C.~decolor)~OTU~1",
                     'C_decolor_2' = "italic(C.~decolor)~OTU~2",
                     'C_impar' = "italic(C.~impar)",
                     'D_polyphaga' = "italic(D.~polyphaga)",
                     'M_bondari_1' = "italic(M.~bondari)~OTU~1",
                     'M_bondari_2' = "italic(M.~bondari)~OTU~2",
                     'M_ypsilon' = "italic(M.~ypsilon)",
                     'P_cocoseae' = "italic(P.~cocoseae)",
                     'R_rectinasus_1' = "italic(R.~rectinasus)",
                     'Syagrus_coronata' = "italic(Syagrus~coronata)",
                     'Syagrus_botryophora' = "italic(Syagrus~botryophora)"
                     )


```


Now let's plot all PCA and save as a supplementary figure
```{r}
p = ggplot(PCA_df) +
  geom_point(aes(x = PC1, y = PC2, color = cluster)) +
  facet_wrap(~taxon,
             scales = 'free',
             labeller = as_labeller(names_translation, default = label_parsed),
             ncol=3
             ) +
  scale_color_brewer(type='qual', guide = 'none',palette = 'Set2') +
  theme_tufte() +
  theme(axis.title = element_blank(),
        panel.border = element_rect(color='black',fill=NA))

print(p)
```

# Mapping PCA axes

Now lets plot PCAs in a map. First, we need to get coordinates for our samples. 

We will also keep the single record of M. bondari in Syagrus coronata.

In this case, since it is a single specimen, we do not have a dataset, but we would still like to map it.

```{r warning=FALSE, message=FALSE}
PCA_map_df = readr::read_csv('sample_info/sample_data.csv') %>%
  dplyr::select(sample=genomics_id,
         plant_species,
         lat,
         lon) %>%
  distinct() %>% #for some reason, a few records are duplicated
  right_join(PCA_df %>% bind_rows(tibble(taxon = 'M_bondari_2',
                                         sample = 'BdM1049',
                                         cluster = '1',
                                         PC1 = 0,
                                         PC2 = 0))
             )

PCA_map_df 
```

Now let's convert the first two PCA axis to the lab color space. First, let's define a function to do it:
```{r}
rgb_cols = function(df){
  x = scales::rescale(df$PC1, from = range(df[c('PC1','PC2')]),  to = c(-100,100))
  y = scales::rescale(df$PC2, from = range(df[c('PC1','PC2')]),  to = c(-100,100))
  l = sqrt(x^2+y^2)/sqrt(2*100^2)*40+50
  
  data.frame(L = l,
             U = y,                                                                       
             V = x
  ) %>%
    convertColor(., from = 'Luv', to = 'sRGB') %>%
    rgb() %>%
    data.frame(sample = df$sample , color=.)
}
```

Now apply the function. We will do a trick to paint M_bondari_2 black:

```{r warning=FALSE, message=FALSE}
PCA_map_df = PCA_map_df %>% 
  group_by(taxon) %>%
  do(rgb_cols(.)) %>%
  ungroup %>%
  dplyr::select(sample, color) %>%
  right_join(PCA_map_df)


color_scale = as.character(PCA_map_df$color)
names(color_scale) = 1:length(PCA_map_df$color)

color_scale['black'] = '#000000'

PCA_map_df$color = as.character(1:length(PCA_map_df$color))
PCA_map_df$color[PCA_map_df$sample == 'BdM1049'] = 'black'

PCA_map_df

```


Now, let's download a base map from Natural Earth (commented ou if already done)
```{r}
basemap = ne_download(scale = 'large', type = 'GRAY_HR_SR_OB_DR', category = 'raster', destdir = './map')
```

After map is downloaded, let's crop it to the region of interest:
```{r}
basemap = raster::stack('map/GRAY_HR_SR_OB_DR/GRAY_HR_SR_OB_DR.tif')
bbox = raster::extent(c(xmin=-44.5,xmax=-35,ymin=-20,ymax=-8))
cropped = raster::crop(basemap,bbox)
map_layer = ggRGB(cropped, ggLayer = TRUE,r = 1,g=1,b=1)
#the following is just to visualize the layer
ggRGB(cropped, r = 1,g=1,b=1) %>%
  print
```


And plot:
```{r}
ggplot(PCA_map_df) +
  map_layer +
  geom_jitter(aes(x = lon, 
                  y = lat, 
                  fill = color,
                  shape = cluster), 
              width=0.2,
              height=0.2, 
              size = 3,
              stroke=0.1) + 
  scale_fill_manual(values = color_scale, guide = F) +
  scale_shape_manual(values = 21:23, guide = F) +
  facet_wrap(~taxon,
             labeller = as_labeller(names_translation, default = label_parsed),
             ncol=5) +
  theme_tufte() +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        strip.text = element_text(size=6)) +
  coord_equal()
```

Now we will do a similar plot, but with plant species as shapes. 


```{r}

names_translation2 = c('Anchylorhynchus_1' = "italic(Anc.~trapez.)~OTU~1",
                     'Anchylorhynchus_2' = "italic(Anc.~trapez.)~OTU~2",
                     'Anchylorhynchus_3' = "italic(Anc.~trapez.)~OTU~3",
                     'Andranthobius_1' = "italic(And.~bondari)~OTU~1",
                     'Andranthobius_2' = "italic(And.~bondari)~OTU~2",
                     'C_decolor_1' = "italic(C.~decolor)~OTU~1",
                     'C_decolor_2' = "italic(C.~decolor)~OTU~2",
                     'C_impar' = "italic(C.~impar)",
                     'D_polyphaga' = "italic(D.~polyphaga)",
                     'M_bondari_1' = "italic(M.~bondari)~OTU~1",
                     'M_bondari_2' = "italic(M.~bondari)~OTU~2",
                     'M_ypsilon' = "italic(M.~ypsilon)",
                     'P_cocoseae' = "italic(P.~cocoseae)",
                     'R_rectinasus_1' = "italic(R.~rectinasus)"
                     )

p = PCA_map_df %>%
  filter(taxon %in% c('Anchylorhynchus_1','Anchylorhynchus_2','Anchylorhynchus_3','M_bondari_1', 'M_bondari_2', 'R_rectinasus_1','M_ypsilon')) %>%
  arrange(plant_species,abs(PC1)*abs(PC2)) %>%
  ggplot +
  map_layer +
  geom_jitter(aes(x = lon, 
                  y = lat, 
                  fill = color,
                  shape = plant_species), 
              width=0.3,
              height=0.3, 
              size = 2,
              stroke=0.05) + 
  scale_fill_manual(values = color_scale, guide = F) +
  scale_shape_manual(values = 21:23, guide = F) +
  facet_wrap(~taxon,
             labeller = as_labeller(names_translation2, default = label_parsed),
             ncol=7) +
  theme_tufte() +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        strip.text = element_text(size=6)) +
  coord_equal(xlim = c(bbox@xmin,bbox@xmax),
              ylim = c(bbox@ymin,bbox@ymax))

print(p)

ggsave(filename = 'figures/PCA_map_pollant.pdf',
       plot = p,
       device = 'pdf',
       height = 4.5,
       width=17.8,
       units = 'cm',
       useDingbats=F)


p = PCA_map_df %>%
  filter(taxon %in% c('Andranthobius_1', 'Andranthobius_2', 'C_decolor_1','C_decolor_2','C_impar','D_polyphaga','P_cocoseae')) %>%
  arrange(plant_species,abs(PC1)*abs(PC2)) %>%
  ggplot +
  map_layer +
  geom_jitter(aes(x = lon, 
                  y = lat, 
                  fill = color,
                  shape = plant_species), 
              width=0.3,
              height=0.3, 
              size = 2,
              stroke=0.05) + 
  scale_fill_manual(values = color_scale, guide = F) +
  scale_shape_manual(values = 21:23, guide = F) +
  facet_wrap(~taxon,
             labeller = as_labeller(names_translation2, default = label_parsed),
             ncol=7) +
  theme_tufte() +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        strip.text = element_text(size=6)) +
  coord_equal(xlim = c(bbox@xmin,bbox@xmax),
              ylim = c(bbox@ymin,bbox@ymax))

print(p)

ggsave(filename = 'figures/PCA_map_commensals.pdf',
       plot = p,
       device = 'pdf',
       height = 4.5,
       width=17.8,
       units = 'cm',
       useDingbats=F)



```

We will manually insert some information in this graph using Adobe Illustrator:
* variance explained by PC axes
* spatial extent of k-means clusters
* an inset with color space mapped on PC axes

The following are blocks of code to generate this information:

Starting by variance explained:
```{r}

PC_text = plyr::llply(eig,
            function(x){
              perc = x$values/sum(x$values)
              paste(c('PC1: ','PC2: '),100*perc[1:2] %>% round(digits = 3),'%\n',sep='',collapse='') %>%
                paste('\n',sep='')
            })

PC_text %>% plyr::l_ply(cat)
```

Now let's list populations for each k-means cluster:
```{r}
PCA_map_df %>%
  mutate(cluster = LETTERS[as.integer(cluster)]) %>%
  left_join(readr::read_csv('sample_info/sample_data.csv') %>%
              dplyr::select(sample=genomics_id,population=pop_id_thesis)) %>%
  group_by(taxon,cluster) %>%
  summarise(pops=paste(unique(population),collapse=','))
  
```

For reference, these are the population coordinates:

```{r}
pop_locs = readr::read_csv('sample_info/sample_data.csv') %>%
              dplyr::select(population=pop_id_thesis, lat, lon) %>%
  group_by(population) %>%
  summarize_all(mean)

ggplot(pop_locs, aes(x=lon,y=lat)) +
  map_layer +
  geom_text(aes(label=population)) +
  coord_equal()
  

  

```



Let's also plot a graph that will work as a guide:

```{r}

key_df = tibble(PC1=seq(-50,50),PC2=seq(-50,50)) %>%
  #filter(Axis1 != 0 & Axis2 != 0) %>%
  expand(crossing(PC1,PC2)) %>%
  mutate(sample=1:dim(.)[1])



key_df = key_df %>%
  left_join(rgb_cols(key_df)) 

colors = unique(as.character(key_df$color))
names(colors) = colors



p = ggplot(key_df, aes(PC1,PC2)) +
  geom_raster(aes(fill=as.character(color)), interpolate = TRUE) +
  geom_hline(aes(yintercept=0)) +
  geom_vline(aes(xintercept=0)) +
  scale_fill_manual(values=colors, guide='none') +
  scale_x_continuous(expand=c(0,0)) + 
  scale_y_continuous(expand=c(0,0)) + 
  theme(plot.margin = margin(1,1,1,1,unit = 'mm'),
        axis.title = element_blank(),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        panel.border = element_rect(fill=NA,colour = 'black',size=1)) +
  coord_cartesian()

print(p)

ggsave(filename = 'figures/PCA_map_guide.pdf',
       plot = p,
       device = 'pdf',
       width = 80,
       height = 60,
       units = 'mm',
       useDingbats=F)

```

# Plant PCA

Our map will also include the range of each plant species and a PCA with their genetic variation.

Let's load plant data, do the PCA and translate to colors. We will do PCA separately for each species of plant since they are completely isolated.


```{r}
plant_spp = c('coronata','botryophora')
names(plant_spp) = str_c('Syagrus_',plant_spp)

plant_specimens = plant_spp %>%
  sapply(function(x){
    list.files(file.path('..','ANGSD_plants',x), 
               pattern = 'bam.filelist',
               recursive = T,
               full.names = T) %>%
      file() %>%
      readLines() %>%
      basename() %>%
      str_remove_all('\\..+$')
  }) %>%
  unlist

plant_sample_names = tibble(sample = plant_specimens, 
                      taxon = names(plant_specimens) %>% str_remove('[0-9]+$'))
plant_sample_names

```


Now let's add covariances to the table.
```{r}
plant_covs = plant_spp %>%
  plyr::llply(function(x){
    list.files(file.path('..','ANGSD_plants',x), 
               pattern = 'pca\\.cov',
               recursive = T,
               full.names = T) %>%
      file %>%
      readr::read_lines() %>%
      plyr::laply(str_split,'[:space:]',simplify = T) %>%
      as.numeric %>%
      matrix(.,nrow=sqrt(length(.)))
  })

str(plant_covs)
```

Now we will do the PCA based on the covariance matrices.

```{r}
plant_eig = plyr::llply(plant_covs, eigen, symmetric = T)
names(plant_eig) = plant_spp

str(plant_eig)
```

Here we combine PCA results with sample names.
```{r warning=FALSE, message=FALSE}

plant_PCA_df = plant_spp %>%
  plyr::ldply(function(i_sp){
    plant_eig[[i_sp]]$vectors[,1:2] %>%
      as.data.frame %>%
      rename(PC1 = 1, PC2 = 2) %>%
      mutate(PC1 = PC1 * sqrt(plant_eig[[i_sp]]$values[1]),
             PC2 = PC2 * sqrt(plant_eig[[i_sp]]$values[2])) %>%
      bind_cols(filter(plant_sample_names, taxon == str_c('Syagrus_', i_sp))) %>%
      dplyr::select(-taxon) #will be added automatically
  },
  .parallel = FALSE) %>%
  rename(taxon = .id)

plant_PCA_df
```


Let's add now location data.
```{r warning=FALSE, message=FALSE}
plant_PCA_map_df = readr::read_csv('sample_info/Syagrus_data.csv') %>%
  dplyr::select(sample=genomics,
         plant_species,
         lat,
         lon,
         pop_id_thesis) %>%
  distinct() %>% #for some reason, a few records are duplicated
  right_join(plant_PCA_df)
        

plant_PCA_map_df 
```


Now let's add a column mapping PCA coordinates to colors
```{r}
plant_PCA_map_df = plant_PCA_map_df %>% 
  group_by(taxon) %>%
  do(rgb_cols(.)) %>%
  ungroup %>%
  dplyr::select(sample, color) %>%
  right_join(plant_PCA_map_df)


plant_color_scale = as.character(plant_PCA_map_df$color)
names(plant_color_scale) = 1:length(plant_PCA_map_df$color)


plant_PCA_map_df$color = as.character(1:length(plant_PCA_map_df$color))


plant_PCA_map_df

```

Before mapping, let's have a look at the PCAs:

```{r}
ggplot(plant_PCA_map_df %>% filter(taxon == 'Syagrus_coronata')) +
  geom_point(aes(x=PC1,y=PC2)) +
  geom_text_repel(aes(x=PC1,y=PC2,label=pop_id_thesis))

ggplot(plant_PCA_map_df %>% filter(taxon == 'Syagrus_botryophora')) +
  geom_point(aes(x=PC1,y=PC2)) +
  geom_text_repel(aes(x=PC1,y=PC2,label=pop_id_thesis))
```


Now let's obtain the map of South America where we will place plant samples.

```{r}

bbox = raster::extent(c(xmin=-44,
              xmax=-35,
              ymin=-20,
              ymax=-8.5)) %>%
  as('SpatialPolygons')

crs(bbox) = crs(basemap)

bbox = st_as_sf(bbox)
  
cropped_Brazil = raster::crop(basemap,raster::extent(c(xmin=-83,xmax=-33,ymin=-30,ymax=0)))

map_extent_sf = raster::extent(c(xmin=-83,xmax=-33,ymin=-31,ymax=1)) %>%
  as('SpatialPolygons') %>%
  st_as_sf() %>%
  st_set_crs(as.character(crs(cropped_Brazil)))
  
countries = ne_countries(scale=50, returnclass = 'sf') %>%
  st_transform(as.character(crs(cropped_Brazil))) %>%
  st_intersection(map_extent_sf)

```


Now, let's plot a map including plant samples:

```{r}

p = ggRGB(cropped_Brazil, r = 1,g=1,b=1) +
  geom_sf(color='black',fill=NA, size=0.2, data=countries) +
  #geom_sf(data=bbox,fill=NA) +
  #geom_sf(data=bbox,fill=NA,color='black') +
  coord_sf() +
  theme_few() +
  xlab(NULL) +
  ylab(NULL) +
  scale_x_continuous(expand = c(0,0), 
                     breaks=c(-80,-40),limits = c(-83,-33)) +
  scale_y_continuous(expand = c(0,0), 
                     breaks=c(0,-30),limits = c(-30,0)) +
  geom_jitter(aes(x = lon, 
                  y = lat, 
                  fill = color,
                  shape = plant_species), 
              data = plant_PCA_map_df,
              width=0.3,
              height=0.3, 
              size = 1.5,
              stroke=0.05) + 
  scale_shape_manual(values = 21:23, guide = F) +
  scale_fill_manual(values = plant_color_scale, guide=F) +
  ggtitle('Syagrus') +
  theme(text = element_text(size=6,face = 'italic'),
        panel.grid = element_blank(),
        plot.background = element_rect(fill = "transparent", color = NA),
        plot.margin = margin(t = 5.5,
                             r = 5.5,
                             b = 5.5,
                             l = 10))

p

ggsave(filename = 'figures/PCA_map_Syagrus.pdf',
       plot = p,
       device = 'pdf',
       width = 80,
       height = 45,
       units = 'mm',
       useDingbats=F)

```


Finally, let's also split plants by kmeans and make a single PCA plot with all species, plants ants insects.

```{r}
set.seed(15897)
plant_clusters = plant_spp %>%
  plyr::llply(function(x){
    find.clusters.eigen(plant_eig[[x]],
                  pca.select = 'percVar',
                  perc.pca=90,
                  max.n.clust = min(10, length(plant_eig[[x]]$values)%/%2))
  },
  .parallel = FALSE)

save(plant_clusters,file = 'temporary_data/plant_clusters_DAPC.RData')
```

```{r}
load(file = 'temporary_data/plant_clusters_DAPC.RData')

names(plant_clusters) = names(plant_covs)
names(plant_eig) = names(plant_covs)
plant_spp = names(plant_covs)
names(plant_spp) = plant_spp

plant_PCA_df = plant_spp %>%
  plyr::ldply(function(i_sp){
    PCA_axes = plant_eig[[i_sp]]$vectors[,1:2] %>%
      as.data.frame %>%
      rename(PC1 = 1, PC2 = 2) %>%
      mutate(PC1 = PC1 * sqrt(plant_eig[[i_sp]]$values[1]),
             PC2 = PC2 * sqrt(plant_eig[[i_sp]]$values[2]),
             cluster = plant_clusters[[i_sp]]$grp) %>%
      bind_cols(filter(plant_sample_names, taxon == i_sp)) %>%
      dplyr::select(-taxon) #will be added automatically
  },
  .parallel = FALSE) %>%
  rename(taxon = .id)

plant_PCA_df
```


Let's now print a graph with PCAs and k-means clusters for all species to use as supplementary material.

```{r}
p = ggplot(PCA_df %>% bind_rows(plant_PCA_df)) +
  geom_point(aes(x = PC1, y = PC2, color = cluster)) +
  facet_wrap(~taxon,
             scales = 'free',
             labeller = as_labeller(names_translation, default = label_parsed),
             ncol=3
             ) +
  scale_color_brewer(type='qual', guide = 'none',palette = 'Set2') +
  theme_tufte() +
  theme(axis.title = element_blank(),
        panel.border = element_rect(color='black',fill=NA))

print(p)

ggsave(filename = 'figures/supp_PCA.pdf',
       plot = p,
       device = 'pdf',
       width = 168,
       height = 168,
       units = 'mm',
       useDingbats=F)
```

The covariance explained by plant PC axes is:

```{r}
plyr::llply(plant_eig,
            function(x){
              perc = x$values/sum(x$values)
              paste(c('PC1: ','PC2: '),100*perc[1:2] %>% round(digits = 3),'%\n',sep='',collapse='') %>%
                paste('\n',sep='')
            }) %>% plyr::l_ply(cat)
```


# Prepare for fastsimcoal

Now the question is: is there gene flow between clusters? 
We will use fastsimcoal to test that. To help producing the input files needed, let's output a table with the assignment of each sample to a population, following the k-means clustering:

```{r}
tb1 = PCA_df %>%
  transmute(taxon,
            sample,
            cluster = LETTERS[cluster]) %>%
  arrange(taxon,cluster,sample)

tb1

readr::write_csv(tb1, path = 'results/DAPC_clusters.csv')
```
